### PR TITLE
Wip/hughsie/update detach

### PR DESCRIPTION
**Pull request by hughsie**

### CODE REVIEW

File: src/fu-plugin.c [hunk 1]
```diff
@@ -967,6 +967,72 @@ fu_plugin_runner_update_cleanup (FuPlugin *plugin, FuDevice *device, GError **er
...
+	/* optional */
+	g_module_symbol (priv->module, "fu_plugin_update_attach", (gpointer *) &func);
+	if (func == NULL)
+		return TRUE;
+	g_debug ("performing update_attach() on %s", priv->name);
+	if (!func (plugin, device, error)) {
+		g_prefix_error (error, "failed to attach device %s: ", priv->name);
+		return FALSE;
+	}
+	return TRUE;
```
Comment (superm1) [inline src/fu-plugin.c]: This is a lot of boiler plate.  I almost wonder if it's better  to have a separate function
`fu_plugin_runner_call_vfunc `that does all of this but takes a 4th `const gchar *` argument for the vfunc being looked up.  The g_debug call can just use that same string as can the error.

Then this function turns into a one liner that looks like:
`return fu_plugin_runner_call_vfunc (plugin, device, error, "fu_plugin_update_attach");`

File: docs/libfwupd/libfwupd-docs.xml [hunk 2]
```diff
@@ -328,6 +328,70 @@ fu_plugin_update_cleanup (FuPlugin *plugin, FuDevice *device, GError **error)
...
+          re-inserting the hardware with a secret button pressed.
+        </para>
+        <para>
+          Before the device update is performed the fwupd daemon runs an optional
+          <code>update_detach()</code> vfunc which switches the device to
+          bootloader mode.
+          After the update (or if the update fails) an the daemon runs an
+          optional <code>update_attach()</code> vfunc which should switch the
+          hardware back to runtime mode.
+          Finally an optional <code>update_coldplug()</code> vfunc is run to
```
[the remark this answers is not in the document]
Comment (hughsie) [reply docs/libfwupd/libfwupd-docs.xml]: Yes, good point, I wasn't 100% happy with the name either. It's really asking the hardware "please refresh *all* the values on the FuDevice that might have changed" so how about `update_refresh()`? Although, that too is somewhat overloaded with the metadata refresh functionality. Maybe `update_reload()`?

File: src/fu-plugin.c (hunk 1 above)
Comment (hughsie) [reply src/fu-plugin.c]: Agreed, fixed.

File: docs/libfwupd/libfwupd-docs.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (superm1) [reply docs/libfwupd/libfwupd-docs.xml]: Well is there a reason that you would ever re-attach and not want to look at versions?  How about just make the requirement that you re-read versions while attaching the new device?  This https://github.com/hughsie/fwupd/pull/312/files#diff-300ca38c94199c9239dc835920e466ecR1311 chunk of code can go away then and you skirt all the potential confusion.

[the remark this answers is not in the document]
Comment (hughsie) [reply docs/libfwupd/libfwupd-docs.xml]: I should have made this a little more clear. I'm working on a patchset that allows two plugins to do half the update each, e.g. a `jabra` plugin could coldplug, and also get the new device firmware version, and the `dfu` plugin could attach and detach. Splitting up "attach" and "reload" is required for that functionality.

[the remark this answers is not in the document]
Comment (superm1) [reply docs/libfwupd/libfwupd-docs.xml]: Ah...
Then yeah I think back to creative name.  not crazy about update_refresh for reasons above.  ```update_reload``` or ```update_reprobe``` both sound good to me.

[the remark this answers is not in the document]
Comment (hughsie) [reply docs/libfwupd/libfwupd-docs.xml]: Changed, thanks.

**Pull request merged**